src/argparse.cpp

```diff
@@ -39,18 +45,31 @@ bool ArgParse::parsingFails(Input& input, Output& output)
     // the number of optional arguments that are provided by 'input':
     size_t optionalArgumentsRemaining = input.size() - mandatoryArguments;
     for (const auto& arg : arguments_) {
-        if (arg.optional_) {
-            if (optionalArgumentsRemaining) {
-                optionalArgumentsRemaining--;
-            } else {
-                // skip this argument if it's optional and there
-                // are not more optional arguments in the input
-                continue;
-            }
+        if (arg.optional_ && optionalArgumentsRemaining == 0) {
+            continue;
         }
         string valueString;
-        // the following can't fail because we checked input.size() before
         input >> valueString;
+        // try to parse this token as a flag
+        if (optionalArgumentsRemaining && tryParseFlag(valueString)) {
+            optionalArgumentsRemaining--;
+            // if this token was the last optional argument
+            if (optionalArgumentsRemaining == 0 && arg.optional_) {
+                // then skip this 'arg'
+                continue;
+            }
+            // otherwise, there is room for another optional
+            // argument or 'arg' is mandatory. So get another
+            // token from the input, and parse it to the
+            // current 'arg'
+            input >> valueString;
+        }
+        // in any case, we have this here:
+        // assert (!arg.optional_ || optionalArgumentsRemaining > 0);
```

Comment:
This statement is left commented out only for documentation purpose?

Reply:
Yes. I don't want any crash if the assertion fails.

Reply:
Ok!!